---
title: "Simulated dataset"
output: html_document
---
The frequency of simulated gene presence state
```{R}
library(phangorn)
library(seqinr)
library(scales)
gpa_f<- '../data/pseudo_gpa/gpa.var.fa'
gpa<- seqinr::read.alignment(file= gpa_f, format = 'fasta')
gpa_mat<- sapply(gpa$seq, function(s) unlist(strsplit(unlist(s), split = '')))
gpa_mat<- t(gpa_mat) # rows: strains

gp_freq<- apply(gpa_mat,2, function(c) sum(c == '1'))
hist(gp_freq, main = 'Frequency of simulated gene presence', 
     nclass = 20,include.lowest = T,xlim = c(0, 200), 
     xlab = 'number of strains')
```

Load the trees
```{R}
library(phytools)
library(ape)
library(reshape2)
library(ggplot2)
source('./cutoffs/tree_correctness.R')

modify_tip_lab<- function(tr){
    tr$tip.label<- sapply(tr$tip.label, function(x) sprintf('SE%s', x))
    tr
}

cutoffs<- seq(from= .05, to= 1, by= .1)
cutoffs<- c(cutoffs, 0.8, 0.7, 0.2, 0.3)
## read trees
tr_files<- list()
ref_tr_f<- '../data/outbreak_sim/phyloFromPtree.nwk'
ref_tr<- modify_tip_lab(read.newick(ref_tr_f))
nuc_tr_f<- '../results/nuc_tr.nwk'
nuc_tr<- modify_tip_lab(read.newick(nuc_tr_f))
for (c in cutoffs){
    col_nuc_tr_f<- sprintf('../results/ll_cutoff/%s/raxml_input/nuc_col.nwk', 
                         as.character(c))
    print(file.exists(col_nuc_tr_f))
    cd_tr_f<- sprintf('../results/ll_cutoff/%s/RAxML_bestTree.conc',
                    as.character(c))
    print(file.exists(cd_tr_f))
    tr_files[[as.character(c)]]<- list('col'= col_nuc_tr_f, 'cd'= cd_tr_f)
    
}
trs<- lapply(tr_files, function(pair) list(col= modify_tip_lab(read.newick(pair$col)), 
                                           cd= modify_tip_lab(read.newick(pair$cd))))
# rooting
outgroup<- c('SE22', 'SE15', 'SE47', 'SE46', 'SE48', 'SE103')
nuc_tr<- root(midpoint.root(nuc_tr), outgroup = outgroup)
trs<- lapply(trs, function(tr_set)
    lapply(tr_set, function(t) root(t, outgroup = outgroup, resolve.root = T)))
```
Compute the topological difference
```{R}
col_rtr_correctness<- sapply(trs, function(pair) tree_correctness(target_tr = pair$col, ref = ref_tr))
cd_rtr_correctness<- sapply(trs, function(pair) tree_correctness(target_tr = pair$cd, ref = ref_tr))
r_baseline_correctness<- tree_correctness(nuc_tr, ref_tr)

col_tr_correctness<- col_rtr_correctness
cd_tr_correctness<- cd_rtr_correctness
baseline_correctness<- r_baseline_correctness

# visualize the results
col_tr_correctness_df<- melt(col_tr_correctness)
col_tr_correctness_df$tr.name<- 'col'
cd_tr_correctness_df<- melt(cd_tr_correctness)
cd_tr_correctness_df$tr.name<- 'cd'

tr_correctness_df<- rbind(col_tr_correctness_df, cd_tr_correctness_df)
colnames(tr_correctness_df)<- c('metric', 'cutoff', 'value', 'tree')
tr_correctness_df$col.perc<- (1-tr_correctness_df$cutoff)*100
ggplot(tr_correctness_df[tr_correctness_df$metric %in% c('TP', 'precision', 'f1'), ])+
    geom_point(aes(x= col.perc, y= value, shape= tree), alpha= .7)+
    geom_line(aes(x= col.perc, y= value, color= tree), alpha= .5)+
    geom_hline(aes(yintercept = baseline_correctness[metric]), linetype= 2, color= 'grey50')+
    xlab('log-likelihood cutoffs (collapsing percentage)')+
    scale_color_manual(values= hue_pal()(2), name= 'tree type', breaks = c('cd', 'col'),
                       labels= c('col'= 'collapsed nucleotide', 'cd'= 'composite datatype'))+
    theme_classic()+
    guides(shape=FALSE)+
    facet_grid(metric~., scale= 'free')
```
Apply different cutoffs to the P. aeruginosa tree
```{R}
setwd('../../Pseudomonas/bin')
library(ggpubr)
source('load_biological_data.R')
# load the trees
tr_files<- list('nuc'= '../results/nuc_tr.nwk', 
                'col25'= '../results/cd_tr.nwk', 
                'col50'= '../results/cd_cutoff50/RAxML_bestTree.conc', 
                'col75'= '../results/cd_cutoff25/RAxML_bestTree.conc', 
                'col100'= '../results/cd_cutoff00/RAxML_bestTree.conc')
#' ensure the filepaths
stopifnot(all(sapply(tr_files, function(f) file.exists(f))))
#' load the trees
trs<- lapply(tr_files, function(f) read.newick(f))
outgroup<- c('CH4433')
trs<- lapply(trs, function(t) root(t, outgroup = outgroup, resolve.root = T))

#' a tree with tips shuffled
set.seed(1234)
rand_tr<-trs$nuc
rand_tr$tip.label<- sample(rand_tr$tip.label)
trs[['rand']]<- rand_tr

sapply(trs, function(t) treedist(t, trs$nuc))

#' load the traits (locations)
bio_df<- load_biological_data()
bio_list<- as.list(bio_df)
for (n in 1:length(bio_list)){
    names(bio_list[[n]])<- rownames(bio_df)
}
traits<- bio_df[trs[['nuc']]$tip.label, 'Origin']
names(traits)<- trs[['nuc']]$tip.label
traits<- factor(traits, levels= c('Charité Berlin', 'University hospital Frankfurt', 'Hannover Medical School'))

mat2df<- function(m){
    df<- melt(m, stringAsFactor= F)
    df[, 1]<- as.character(df[, 1])
    df[, 2]<- as.character(df[, 2])
    df[df[, 1] > df[, 2], ]
}
conc_locs<- function(locs){
    ordered_locs<- locs[order(locs)]
    paste0(ordered_locs, collapse = '-')
}
# significance of change of cophenetic distances
determine_sig<- function(count_df, h1, h2, alternative= NA){
    print(sprintf('%s-%s', h1, h2))
    if (is.na(alternative)){
        alternative<-'less'
        if (h1 == h2){
            alternative= 'greater'
        }    
    }
    
    sig<- Inf 
    if (nrow(count_df[(count_df$loc1 == h1) & (count_df$loc2 == h2), ]) > 0){
        t_out<- t.test(count_df[(count_df$loc1 == h1) & (count_df$loc2 == h2), 1], 
                       count_df[(count_df$loc1 == h1) & (count_df$loc2 == h2), 2], 
                       paired = TRUE, alternative = alternative)      
        sig<- t_out$p.value
    }
    sig
}
#' cophenetic distances (number of edges between tips)
cophenetic_plots<- list()
plot_max_val<- 15
for (tr.name1 in c('col25', 'col50', 'col75', 'col100')){
    tr.name2<- 'nuc'
    violin_p<- list()
    bars_p<- list()
    cophe_p<- list()
    
    # calculate cophenetic distances
    cophe_mats<- lapply(list('cd'= trs[[tr.name1]], 'nuc'= trs[[tr.name2]]), function(t)cophenetic.phylo(compute.brlen(t, 1)))
    cophe_dfs<- lapply(cophe_mats, function(m) mat2df(m))
    cophe_one_df<- merge(cophe_dfs$nuc, cophe_dfs$cd, by=c('Var1', 'Var2')) 
    colnames(cophe_one_df)<- c('sample1', 'sample2', 'nuc_dist', 'cd_dist')
    cophe_one_df$loc1<-  as.character(traits[cophe_one_df$sample1])
    cophe_one_df$loc2<-  as.character(traits[cophe_one_df$sample2])
    cophe_one_df$loc_pair<- apply(cophe_one_df, 1, function(r) conc_locs(c(r['loc1'], r['loc2'])))
    
    
    
    loc1s<- c()
    loc2s<- c()
    sigs<- c()
    alternative_hs<- c()
    pairs_num<- c()
    sub_cophe_one_df<- data.frame()
    for (h1 in unique(traits)){
        for (h2 in unique(traits)){
            loc_filter<- (cophe_one_df$loc1 == h1) & (cophe_one_df$loc2 == h2)
            if (sum(loc_filter) == 0){
                next
            }
            
            count_df<- cophe_one_df[loc_filter, 
                                    c('nuc_dist', 'cd_dist', 'loc1', 'loc2')]
            sub_cophe_one_df<- rbind(sub_cophe_one_df, 
                                     cophe_one_df[loc_filter & 
                                                      (cophe_one_df$nuc_dist != cophe_one_df$cd_dist),] )
            print(table(count_df$loc1, count_df$loc2))
            loc1s<- c(loc1s, h1)
            loc2s<- c(loc2s, h2)
            sigs<- c(sigs, determine_sig(count_df, h1, h2))
            pairs_num<- c(pairs_num, nrow(count_df))
            print(determine_sig(count_df, h1, h2))
            alternative_hs<-c(alternative_hs, ifelse(h1==h2, 'greater', 'less'))
        }
    }
    
    data.frame(loc1s, loc2s, sigs, alternative_hs, pairs_num)
    sub_cophe_one_df$loc_pair<-factor(sub_cophe_one_df$loc_pair, 
                                      levels= c('Charité Berlin-Charité Berlin', 
                                                'University hospital Frankfurt-University hospital Frankfurt',
                                                'Hannover Medical School-Hannover Medical School',
                                                'Charité Berlin-University hospital Frankfurt',
                                                'Charité Berlin-Hannover Medical School',
                                                'Hannover Medical School-University hospital Frankfurt'))
    violin_p<- ggplot(sub_cophe_one_df, aes(x= loc_pair))+
        geom_violin(aes(y= cd_dist-nuc_dist, fill= loc1==loc2))+
        scale_y_continuous(limits= c(-plot_max_val, plot_max_val), 
                           breaks= seq(from= -plot_max_val,to= plot_max_val,by=3))+
        theme_classic()+
        ylab('distance difference (cd - nuc')+
        theme(axis.text.x = element_text(angle = 60, hjust = 1))
    
    # count the percentages for each group
    change_cats<- ifelse(sub_cophe_one_df$cd_dist > sub_cophe_one_df$nuc_dist, 'increased', 'decreased')
    sub_cophe_one_stat_df<- dcast(
        as.data.frame(
            table(change_cats, sub_cophe_one_df$loc_pair)), Var2~change_cats)
    rownames(sub_cophe_one_stat_df)<- sub_cophe_one_stat_df[, 1]
    sub_cophe_one_stat_df[, 1]<- NULL
    sub_cophe_one_perc_df<- sub_cophe_one_stat_df/rowSums(sub_cophe_one_stat_df) * 100
    colnames(sub_cophe_one_perc_df)<- colnames(sub_cophe_one_stat_df)
    
    sub_cophe_one_stat_df$hospitals<- rownames(sub_cophe_one_stat_df)
    sub_cophe_one_perc_df$hospitals<- rownames(sub_cophe_one_perc_df)
    sub_cophe_one_stat_long<- melt(sub_cophe_one_stat_df)
    sub_cophe_one_stat_long<- sub_cophe_one_stat_long[order(sub_cophe_one_stat_long$hospitals, sub_cophe_one_stat_long$variable), ]
    sub_cophe_one_perc_long<- melt(sub_cophe_one_perc_df)
    sub_cophe_one_perc_long<- sub_cophe_one_perc_long[order(sub_cophe_one_perc_long$hospitals, sub_cophe_one_perc_long$variable), ]
    
    plot_df<- cbind(sub_cophe_one_stat_long, sub_cophe_one_perc_long$value)
    plot_df$hospitals<- factor(plot_df$hospitals, levels= c('Charité Berlin-Charité Berlin', 
                                                            'University hospital Frankfurt-University hospital Frankfurt',
                                                            'Hannover Medical School-Hannover Medical School',
                                                            'Charité Berlin-Hannover Medical School',
                                                            'Charité Berlin-University hospital Frankfurt',
                                                            'Hannover Medical School-University hospital Frankfurt'))
    colnames(plot_df)<- c('hospitals', 'change', 'count', 'percentage')
    bars_p<- ggplot(plot_df, aes(x= hospitals, y= percentage,fill= change))+
        geom_bar(stat="identity")+
        geom_text(aes(label=count, color= change), position = position_stack(vjust= .5))+
        scale_fill_manual(values=c('decreased'='yellowgreen',  'increased'='dodgerblue4'))+
        scale_color_manual(values= c('decreased'= 'black', 'increased'= 'grey'))+
        theme_classic()+
        theme(axis.text.x = element_text(angle = 60, hjust = 1))
    
    cophenetic_plots[[tr.name1]]<- ggarrange(plotlist = list('violin'= violin_p, 'bars'= bars_p), ncol= 1)
}
# check cophenetic_plots for each application
```
