---
title: "R Notebook"
output: html_notebook
---
__Brief__
Population structure determined using fastBAPS, the Bayesian analysis of population structure. Transmisssion analysis will be done for each identified group.

First determine the groups from nucleotide tree
```{R}
library(fastbaps)
library(ggtree)
library(phytools)
library(ggplot2)
newick.file.name<- '~/bifo-sgi/TreePaper/MRKP.v2/results/after_nuc_tree/v1/nuc.col.nwk'
tr <- phytools::read.newick(newick.file.name)
ggtree(tr, branch.length = 'none')+
    geom_nodelab(aes(label= node))
# sub_tr<- keep.tip(tr, get.offspring.tip(tr, node= 106))
```
Trim the alignment and detect the population structure
```{R,fig.height = 10, fig.width = 5, fig.align = "center"}
library(seqinr)
target_tips<- get.offspring.tip(tr, node= 106)
fasta.file.name<- '~/bifo-sgi/TreePaper/MRKP.v2/results/alignment/nuc.var.aln'
aln<- read.alignment(file= fasta.file.name, 
                     format= 'fasta')
target_seqs<- lapply(which(aln$nam %in% target_tips), function(n) aln$seq[[n]])
target_nams<- lapply(which(aln$nam %in% target_tips), function(n) aln$nam[n])

sub_aln_f<- './nuc.var.target_tips.aln'
write.fasta(sequences = target_seqs, names = target_nams, file.out = sub_aln_f, nbchar = 70)
## run fastBUPS
set.seed(1234)
sparse.data <- import_fasta_sparse_nt(sub_aln_f)
sub_tr<- keep.tip(tr, target_tips)
best.partition<- best_baps_partition(sparse.data, sub_tr)
plot.df <- data.frame(id = sub_tr$tip.label, 
                      fastbaps = as.character(best.partition), 
                      stringsAsFactors = FALSE)
gg<- ggtree(sub_tr)+
    geom_tiplab()+
    geom_treescale(x= 0, offset= -2)
f2 <- facet_plot(gg, panel = "fastbaps", data = plot.df, geom = geom_tile, 
                 aes(x= 1, fill = fastbaps),color= 'white')+
    scale_fill_manual(values= c('#188C77', '#F2AE2E', '#262626'))
     xlim_tree(1e-2)
f2
```
Then, turn to the composite datatype tree
```{R}
groups<- lapply(1:length(unique(best.partition)), function(x) names(best.partition)[which(best.partition == x)])
cd_tr_f<- '~/bifo-sgi/TreePaper/MRKP.v2/results/after_nuc_tree/v1/md.col.nwk'
cd_tr<- read.newick(cd_tr_f)
lapply(groups, function (g) is.monophyletic(phy = cd_tr, tips = g))
```
Prune the trees 
```{R}
all_subtrees<-list()
all_subtrees$nuc<- lapply(groups, function(g) keep.tip(phy = tr, tip = g))
all_subtrees$cd<- lapply(groups, function(g) keep.tip(phy = cd_tr, tip = g))
## save the trees for dating and transmission analysis
# d<- '~/bifo-sgi/TreePaper/MRKP.v2/results/after_nuc_tree/v3/beast'
# for (x in names(all_subtrees)){
#     for (group_name in names(all_subtrees[[x]])){
#         
#         output_d<- paste(c(d, x, group_name), collapse = '/')
#         output_f<- paste(c(output_d, 'pruned_tr.nwk'), collapse = '/')
#         dir.create(output_d, recursive = T)
#         write.tree(phy = all_subtrees[[x]][[group_name]], file = output_f)
#     }
# }

